### PR TITLE
feat(python): expand NumPy input support across the API

### PR DESCRIPTION
**Pull request by pinkfish**
## Summary

- Build on #891's vector support by accepting NumPy arrays, tuples, and other valid sequences wherever the Python API accepts coordinate or index lists.
- Replace dtype- and layout-specific NumPy buffer parsing with Python's generic sequence protocol, without adding a build-time NumPy dependency.
- Cover matrices, polygons and polyhedra, polylines and splines, extrusions, transforms, imports, analysis functions, and integer index data.
- Add PythonSCAD-only `Vector1`, `Vector2`, `Vector3`, and `Matrix4x4` helpers, using NumPy when available and list-backed fallbacks otherwise.
- Update type stubs and add conversion, helper-type, and embedded-interpreter regression coverage, including environments where NumPy is optional.

## Relationship to #891

#891 added NumPy buffer handling to `python_vectorval()` and `python_vectors()`. This PR supersedes that implementation with generic sequence conversion, fixes scalar integer and buffer-layout edge cases, and extends support to API entry points and converters that #891 did not cover.

## Test plan

- Exercise vector and matrix helpers with both NumPy-backed and list-backed implementations.
- Compare list, tuple, and NumPy inputs across Python conversion APIs.
- Run embedded echo coverage with and without NumPy available.

### CODE REVIEW

File: libraries/python/pythonscad/_vectors.py
Comment (nomike) [inline libraries/python/pythonscad/_vectors.py:1]: IMHO this should go into the `pythonscad` package and here's why:

In the past we only had the `openscad` package, which was the C-API implementation. As we wanted to be able to also write pure Python functions and classes, we renamed the C-API module to `_openscad` which is then inherited by the `openscad`.

Way before that change (and even before I joined the project), OpenSCAD merged parts of our Python functionality.

I have thus further split the module into `openscad` and `pythonscad`, with the idea that stuff which is only available in PythonSCAD going to the `pythonscad` module and stuff which is also available in OpenSCAD being in the `openscad`  module.

As `pythonscad` inherits from `openscad` this allows us to move stuff from `pythonscad` to `openscad` in the future, in case they ever decide to merge it.

File: tests/test_python_numpy_api.py
Comment (nomike) [inline tests/test_python_numpy_api.py:1]: We generally use a few categories of tests:

There are the echo tests, which produce test output and are matched against a known-good output, we also have image-render tests where a design is rendered and matched against a golden image and we have 3D render tests where an STL or 3MF file is produced and compared against a golden file.

Test files are usually auto-detected by CMake.

This here however, would introduce a separate test which is outside of the existing categories. While it is okay to do so for special cases where nothing else is possible, I believe this one could be handled as a echo-test, where the script does it's tests internally and then either prints "OK" or crashes with a stack trace or some meaningful error message.

I will try to modify the code accordingly.

**Review by nomike — COMMENTED** [review]
The two requested module-layout and test-harness adaptations are implemented in f262a415e.

File: libraries/python/pythonscad/_vectors.py
Comment (nomike) [reply libraries/python/pythonscad/_vectors.py:1]: I moved the runtime helpers to `pythonscad._vectors`, re-export them only from `pythonscad`, and moved their concrete type stubs to `pythonscad-stubs`. The shared C API still accepts NumPy arrays and tuples through `_openscad`, but `openscad` no longer exposes the PythonSCAD-only helper classes.

File: tests/test_python_numpy_api.py
Comment (nomike) [reply tests/test_python_numpy_api.py:1]: I replaced the standalone `python-numpy-api` CMake registration and wrapper with the auto-discovered `pythonscad-overlay-echo/numpy-api.py` fixture and an `OK` golden output. Its assertions still compare list/NumPy geometry and now also verify the module boundary; failures surface as normal tracebacks. The separate host-Python vector unit test remains only because it deterministically blocks NumPy to exercise the list-backed fallback, which an already-running embedded interpreter cannot do.

File: src/python/py_extrude.cc [hunk 1]
```diff
@@ -378,9 +399,10 @@ PyObject *python_path_extrude(PyObject *self, PyObject *args, PyObject *kwargs)
 
   char *kwlist[] = {"obj",   "path",   "xdir", "convexity", "origin", "scale",
                     "twist", "closed", "fn",   "fa",        "fs",     NULL};
-  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "OO!|O!iOOOOOddd", kwlist, &obj, &PyList_Type, &path,
-                                   &PyList_Type, &xdir, &convexity, &origin, &scale, &twist, &closed,
-                                   &allow_intersect, &fn, &fs, &fs)) {
+  // path/xdir accept list/tuple/NumPy; validated by python_vectorval() in
+  // path_extrude_core().
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "OO|OiOOOOOddd", kwlist, &obj, &path, &xdir, &convexity,
+                                   &origin, &scale, &twist, &closed, &allow_intersect, &fn, &fs, &fs)) {
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_extrude.cc]: Fixed both call sites in 109d5ee24: `fa` is now parsed into `fa` rather than passing `fs` twice.

File: src/python/py_primitives.cc [hunk 2]
```diff
@@ -650,52 +658,84 @@ PyObject *python_polyhedron(PyObject *self, PyObject *args, PyObject *kwargs)
...
             return NULL;
           }
           face.push_back(pointIndex);
         }
+        Py_DECREF(fseq);
         if (face.size() >= 3) {
           node->faces.push_back(std::move(face));
         } else {
+          Py_DECREF(seq);
           PyErr_SetString(PyExc_TypeError, "Polyhedron Face must sepcify at least 3 indices");
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_primitives.cc]: Fixed the typo in 109d5ee24.

File: src/python/py_primitives.cc [hunk 3]
```diff
@@ -650,52 +658,84 @@ PyObject *python_polyhedron(PyObject *self, PyObject *args, PyObject *kwargs)
...
+          }
+          long pointIndex = (long)dind;
           if (pointIndex < 0 || pointIndex >= static_cast<long>(node->points.size())) {
+            Py_DECREF(fseq);
+            Py_DECREF(seq);
             PyErr_SetString(PyExc_TypeError, "Polyhedron Point Index out of range");
             return NULL;
           }
           face.push_back(pointIndex);
         }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_primitives.cc:700]: Fixed in 109d5ee24. Face indices now use `PyNumber_Index()`, preserving NumPy integer support while rejecting floating-point values; a regression assertion covers NumPy float faces.

File: src/python/pyconversion.cc [hunk 4]
```diff
@@ -254,65 +287,98 @@ PyObject *python_fromvector(const Vector3d vec)
...
       }
+      long pointIndex = (long)dval;
+      if (pointIndex < 0) {  // TODO fix || pointIndex >= node->points.size()) {
+        PyErr_SetString(PyExc_TypeError, "Polyhedron Point Index out of range");
+        Py_DECREF(sub);
+        Py_DECREF(seq);
+        return result;
+      }
+      path.push_back(pointIndex);
     }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/pyconversion.cc:396]: Fixed in 109d5ee24. Polygon path indices now use the same integer-only `PyNumber_Index()` conversion, with a NumPy float-index regression assertion.

File: src/python/pyconversion.cc [hunk 5]
```diff
@@ -89,108 +98,99 @@ int python_numberval(PyObject *number, double *result, int *flags, int flagor)
...
+      PyErr_Clear();
+      return result;
+    }
+    Py_ssize_t size = PySequence_Fast_GET_SIZE(seq);
+    for (Py_ssize_t i = 0; i < size; i++) {
+      double val;
+      if (!python_numberval(PySequence_Fast_GET_ITEM(seq, i), &val, nullptr, 0)) {
+        result.push_back((int)val);
       }
     }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/pyconversion.cc:159]: Fixed in 109d5ee24. `python_intlistval()` now accepts native and NumPy integer scalars through `PyNumber_Index()` and no longer truncates floats; invalid values retain the pre-PR behavior of being ignored.

File: src/python/pyopenscad.cc [hunk 6]
```diff
@@ -947,6 +962,18 @@ Value python_convertresult(PyObject *arg, int& error)
...
+    // NumPy scalars (numpy.float64, numpy.int64, ...) and other objects
+    // implementing the number protocol collapse to a double.
+    PyObject *f = PyNumber_Float(arg);
+    if (f != nullptr) {
+      double d = PyFloat_AsDouble(f);
+      Py_DECREF(f);
+      return {d};
+    }
+    PyErr_Clear();
+    PyErr_SetString(PyExc_TypeError, "Unsupported numeric function result\n");
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/pyopenscad.cc]: Fixed the inconsistent trailing newline in 109d5ee24.

File: libraries/python/stubs/_openscad-stubs/__init__.pyi [hunk 7]
```diff
@@ -611,7 +654,7 @@ def circle(
     ...
 
 def polygon(
-    points: Matrix4x4, paths: Optional[list[list[int]]] = None, convexity: int = 2
+    points: Matrix4x4, paths:list[list[int]] | Annotated[npt.NDArray[np.float64], Literal["Any", "Any"]] | None = None, convexity: int = 2
 ) -> PyOpenSCAD:
```
[the remark this answers is not in the document]
Comment (nomike) [reply libraries/python/stubs/_openscad-stubs/__init__.pyi:620]: Fixed in 109d5ee24. The stub now describes polygon points as nested numeric sequences or a NumPy coordinate array, and paths as nested integer sequences or a NumPy int64 array.

File: src/python/pyopenscad.cc [hunk 8]
```diff
@@ -686,15 +686,21 @@ Outline2d python_getprofile(void *v_cbfunc, int fn, double arg)
...
+        PyObject *pypt = PySequence_Fast_GET_ITEM(seq, i);
+        double x, y;
+        if (python_vectorval(pypt, 2, 2, &x, &y, nullptr, nullptr, nullptr) == 0) {
+          result.vertices.push_back(Vector2d(x, y));
+        }
       }
+      Py_DECREF(seq);
     }
   }
   if (result.vertices.size() < 3) {
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/pyopenscad.cc:708]: Fixed in d45442e2f by releasing the callback result after sequence processing and before either return path.

File: src/python/py_extrude.cc [hunk 9]
```diff
@@ -348,9 +366,12 @@ PyObject *path_extrude_core(PyObject *obj, PyObject *path, PyObject *xdir, int c
...
   }
 
-  if (scale != NULL && PyList_Check(scale) && PyList_Size(scale) == 2) {
-    node->scale_x = PyFloat_AsDouble(PyList_GetItem(scale, 0));
-    node->scale_y = PyFloat_AsDouble(PyList_GetItem(scale, 1));
+  {
+    double sx, sy;
+    if (scale != NULL && python_vectorval(scale, 2, 2, &sx, &sy, nullptr, nullptr, nullptr) == 0) {
+      node->scale_x = sx;
+      node->scale_y = sy;
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_extrude.cc]: Fixed in d45442e2f by removing the redundant second scale conversion.

File: libraries/python/stubs/_openscad-stubs/__init__.pyi [hunk 10]
```diff
@@ -351,10 +354,10 @@ class PyOpenSCAD:
 
     def path_extrude(
         self,
-        path: List[float],
-        xdir: List[float],
+        path: List[float]| "npt.NDArray[np.float64]",
+        xdir: List[float]| "npt.NDArray[np.float64]",
         convexity: int = 2,
-        origin: List[float] = [0, 0, 0],
+        origin: List[float]| "npt.NDArray[np.float64]" = [0, 0, 0],
```
[the remark this answers is not in the document]
Comment (nomike) [reply libraries/python/stubs/_openscad-stubs/__init__.pyi]: Fixed in d45442e2f. The method stub now models `path` as a nested coordinate sequence/2D array and uses the runtime two-element origin default.

File: libraries/python/stubs/_openscad-stubs/__init__.pyi [hunk 11]
```diff
@@ -1089,10 +1094,10 @@ def rotate_extrude(
 
 def path_extrude(
     obj: PyOpenSCAD,
-    path: List[float],
-    xdir: List[float],
+    path: List[float]| "npt.NDArray[np.float64]",
+    xdir: List[float]| "npt.NDArray[np.float64]",
     convexity: int,
-    origin: List[float],
+    origin: List[float]| "npt.NDArray[np.float64]",
```
[the remark this answers is not in the document]
Comment (nomike) [reply libraries/python/stubs/_openscad-stubs/__init__.pyi]: Fixed in d45442e2f. The module-level stub now models `path` as a nested coordinate sequence or 2D NumPy array.

File: src/python/py_extrude.cc [hunk 12]
```diff
@@ -75,13 +75,17 @@ PyObject *rotate_extrude_core(PyObject *obj, int convexity, double scale, double
...
+    double ox, oy;
+    if (origin != NULL && python_vectorval(origin, 2, 2, &ox, &oy, nullptr, nullptr, nullptr) == 0) {
+      node->origin_x = ox;
+      node->origin_y = oy;
+    }
+    if (offset != NULL && python_vectorval(offset, 2, 2, &ox, &oy, nullptr, nullptr, nullptr) == 0) {
+      node->offset_x = ox;
+      node->offset_y = oy;
+    }
   }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_extrude.cc:97]: Fixed in 78193f3f7. Provided rotate_extrude origin/offset and linear_extrude origin/scale values now propagate conversion failures as TypeError instead of silently retaining defaults; the echo regression test covers each malformed parameter.

File: src/python/pyconversion.cc [hunk 13]
```diff
@@ -373,122 +447,90 @@ PyObject *python_from2dlong(const std::vector<IndexedFace>& intlist)
...
+              for (Py_ssize_t i = 0; i < m && i < 3; i++) {
+                if (python_numberval(PySequence_Fast_GET_ITEM(inner, i), &result[i], nullptr, 0)) {
+                  Py_DECREF(inner);
+                  Py_DECREF(seq);
                   return results;  // Error
+                }
               }
             }
+            Py_DECREF(inner);
           }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/pyconversion.cc:517]: Fixed in 78193f3f7. python_vectors now rejects flat and nested vectors whose dimensions fall outside the requested range instead of producing zero vectors; both cases have regression coverage.

File: src/python/py_primitives.cc [hunk 14]
```diff
@@ -882,8 +923,9 @@ PyObject *python_polygon(PyObject *self, PyObject *args, PyObject *kwargs)
   PyObject *pypaths = NULL;
   int convexity = 2;
 
-  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O!|O!i", kwlist, &PyList_Type, &pypoints, &PyList_Type,
-                                   &pypaths, &convexity)) {
+  // Accept lists, tuples or NumPy arrays; type is validated by
+  // python_to2dvarpointlist() / python_to2dintlist() below.
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O|Oi", kwlist, &pypoints, &pypaths, &convexity)) {
     PyErr_SetString(PyExc_TypeError, "Error during parsing polygon(points,paths)");
     return NULL;
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_primitives.cc:932]: Fixed in 78193f3f7. polygon now checks the Python error indicator after python_to2dintlist() and returns NULL, preserving the original TypeError rather than causing SystemError; an empty-path regression case was added.

File: src/python/py_transform.cc [hunk 15]
```diff
@@ -261,14 +261,11 @@ PyObject *python_rotate_core(PyObject *obj, PyObject *val_a, PyObject *val_v, Py
   Vector3d vec3(0, 0, 0);
   double angle;
   int dragflags = 0;
-  if (val_a != nullptr && PyList_Check(val_a) && val_v == nullptr) {
+  if (val_a != nullptr && python_is_sequence(val_a) && val_v == nullptr) {
     python_vectorval(val_a, 1, 3, &(vec3[0]), &(vec3[1]), &(vec3[2]), nullptr, &dragflags);
     return python_rotate_sub(obj, vec3, NAN, ref, dragflags);
   } else if (val_a != nullptr && val_v != nullptr && !python_numberval(val_a, &angle, nullptr, 0) &&
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_transform.cc:270]: Fixed in 517422ab2. python_rotate_core() now checks python_vectorval() and raises TypeError for malformed rotation vectors; the echo regression test covers an invalid-length vector.

File: src/python/py_extrude.cc [hunk 16]
```diff
@@ -75,13 +75,23 @@ PyObject *rotate_extrude_core(PyObject *obj, int convexity, double scale, double
...
-    node->offset_x = PyFloat_AsDouble(PyList_GetItem(offset, 0));
-    node->offset_y = PyFloat_AsDouble(PyList_GetItem(offset, 1));
+  if (offset != NULL) {
+    if (python_vectorval(offset, 2, 2, &ox, &oy, nullptr, nullptr, nullptr)) {
+      PyErr_SetString(PyExc_TypeError, "Invalid rotate_extrude offset parameter");
+      return NULL;
+    }
+    node->offset_x = ox;
+    node->offset_y = oy;
   }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_extrude.cc:97]: Fixed in 517422ab2. rotate_extrude origin/offset and linear_extrude origin/scale now preserve explicit None as the default sentinel while still rejecting malformed non-None vectors. Both APIs have regression coverage for explicit None.

File: src/python/py_analysis.cc [hunk 17]
```diff
@@ -244,7 +244,9 @@ PyObject *python_bbox_core(PyObject *obj)
   // Create bounding shape: square for 2D, cube for 3D
   PyObject *bbox_args = PyTuple_New(0);
   PyObject *bbox_kwargs = PyDict_New();
-  bool is_2d = PyList_Check(size) && PyList_Size(size) == 2;
+  // `size` may be a plain list or a NumPy array (when numpy is installed),
+  // so probe it through the sequence protocol rather than PyList_*.
+  bool is_2d = python_is_sequence(size) && PySequence_Size(size) == 2;
   PyDict_SetItemString(bbox_kwargs, is_2d ? "dim" : "size", size);
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_analysis.cc]: Fixed in 517422ab2. python_bbox_core() now handles a negative PySequence_Size() result immediately, releases the owned position/size references, and propagates the original Python exception.

File: src/python/py_csg.cc [hunk 18]
```diff
@@ -404,12 +404,20 @@ PyObject *python_nb_sub_vec3(PyObject *arg1, PyObject *arg2,
   std::vector<Vector3d> vecs;
   int dragflags = 0;
   if (mode == 3) {
-    if (!PyList_Check(arg2)) {
+    // Accept a list/tuple/NumPy array for the explode spec.
+    if (!python_is_sequence(arg2)) {
       PyErr_SetString(PyExc_TypeError, "explode arg must be a list");
       return NULL;
     }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_csg.cc:411]: Fixed in e5434e27f. Explode validation errors now consistently describe the accepted input as a sequence and use clearer maximum-direction wording.

File: libraries/python/stubs/pythonscad-stubs/__init__.pyi [hunk 19]
```diff
@@ -15,15 +15,47 @@ import typing as _typing
...
+class Vector1(_VectorBase):
+    """1D vector represented as [x]."""
+
+class Vector2(_VectorBase):
+    """2D vector represented as [x, y]."""
+
+class Vector3(_VectorBase):
+    """3D vector represented as [x, y, z]."""
+
+class Matrix4x4(list[list[float]]):
```
[the remark this answers is not in the document]
Comment (nomike) [reply libraries/python/stubs/pythonscad-stubs/__init__.pyi]: Fixed in e5434e27f. The vector and matrix stubs now inherit from typed NumPy ndarray bases, exposing ndarray behavior such as shape and arithmetic and avoiding the incorrect list-subclass model used by type-aware tooling.

File: src/python/pyconversion.cc [hunk 20]
```diff
@@ -208,19 +223,40 @@ PyObject *python_frommatrix(const Matrix4d& mat)
...
+    for (Py_ssize_t j = 0; j < std::min((Py_ssize_t)4, ncols); j++) {
+      double val;
+      if (python_numberval(PySequence_Fast_GET_ITEM(cells, j), &val, nullptr, 0)) {
+        rc = 1;
+        break;
+      }
       mat(i, j) = val;
     }
+    Py_DECREF(cells);
   }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/pyconversion.cc:266]: Fixed in 656c5b9c5. python_tomatrix() now requires exactly four rows and four columns per row; a partial-matrix regression case verifies multmatrix raises TypeError.

File: libraries/python/stubs/pythonscad-stubs/__init__.pyi [hunk 21]
```diff
@@ -12,18 +12,51 @@ re-exports `_openscad`). PythonSCAD-only additions are surfaced here.
 # at runtime.
 import typing as _typing
 
+import numpy as _np
 from openscad import *  # noqa: F401,F403
```
[the remark this answers is not in the document]
Comment (nomike) [reply libraries/python/stubs/pythonscad-stubs/__init__.pyi:19]: Fixed in 656c5b9c5. The stub now imports NumPy best-effort and falls back to Any when unavailable, preserving optional NumPy environments while exposing ndarray behavior when type information is installed.

File: src/python/py_io.cc [hunk 22]
```diff
@@ -506,17 +507,14 @@ PyObject *do_import_python(PyObject *self, PyObject *args, PyObject *kwargs, Imp
...
   double originY = 0;
   if (origin != nullptr && origin != Py_None) {
-    originX = PyFloat_AsDouble(PyList_GetItem(origin, 0));
-    if (PyErr_Occurred()) return nullptr;
-    originY = PyFloat_AsDouble(PyList_GetItem(origin, 1));
-    if (PyErr_Occurred()) return nullptr;
+    if (python_vectorval(origin, 2, 2, &originX, &originY, nullptr, nullptr, nullptr) != 0) {
+      PyErr_SetString(PyExc_TypeError,
+                      "osimport(): origin must be a two-element coordinate sequence or None");
+      return nullptr;
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_io.cc:517]: Fixed in 221aa7bf7. osimport origin now requires a sequence before vector conversion, so scalar broadcasting cannot bypass the documented coordinate contract; regression coverage was added.

File: src/python/py_extrude.cc [hunk 23]
```diff
@@ -75,13 +75,23 @@ PyObject *rotate_extrude_core(PyObject *obj, int convexity, double scale, double
...
+  // origin/offset accept a list/tuple/NumPy 2-vector.
+  double ox, oy;
+  if (origin != NULL && origin != Py_None) {
+    if (python_vectorval(origin, 2, 2, &ox, &oy, nullptr, nullptr, nullptr)) {
+      PyErr_SetString(PyExc_TypeError, "Invalid rotate_extrude origin parameter");
+      return NULL;
+    }
+    node->origin_x = ox;
+    node->origin_y = oy;
   }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_extrude.cc:88]: Fixed in 221aa7bf7. rotate_extrude now rejects scalar origin and offset values before conversion, with regression coverage for both parameters.

File: src/python/py_extrude.cc [hunk 24]
```diff
@@ -183,16 +193,26 @@ PyObject *linear_extrude_core(PyObject *obj, PyObject *height, int convexity, Py
...
-    node->origin_y = PyFloat_AsDouble(PyList_GetItem(origin, 1));
+  if (origin != NULL && origin != Py_None) {
+    double ox, oy;
+    if (python_vectorval(origin, 2, 2, &ox, &oy, nullptr, nullptr, nullptr)) {
+      PyErr_SetString(PyExc_TypeError, "Invalid linear_extrude origin parameter");
+      return NULL;
+    }
+    node->origin_x = ox;
+    node->origin_y = oy;
   }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_extrude.cc:207]: Fixed in 221aa7bf7. linear_extrude now rejects scalar origin and scale values while continuing to accept explicit None sentinels; both cases are covered.

File: src/python/py_primitives.cc [hunk 25]
```diff
@@ -614,31 +614,39 @@ PyObject *python_polyhedron(PyObject *self, PyObject *args, PyObject *kwargs)
...
-      } else {
+    for (Py_ssize_t i = 0; i < npoints; i++) {
+      element = PySequence_Fast_GET_ITEM(seq, i);
+      if (python_vectorval(element, 3, 3, &point[0], &point[1], &point[2], nullptr, nullptr)) {
+        Py_DECREF(seq);
         PyErr_SetString(PyExc_TypeError, "Coordinate must exactly contain 3 numbers");
         return NULL;
       }
+      node->points.push_back(point);
     }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/py_primitives.cc:649]: Fixed in 221aa7bf7. Polyhedron points and face colors now require each entry to be a coordinate sequence before conversion. The same validation was applied to polygon, spline, and path_extrude point lists, with regression coverage.

File: src/python/pyopenscad.cc [hunk 26]
```diff
@@ -686,17 +686,24 @@ Outline2d python_getprofile(void *v_cbfunc, int fn, double arg)
...
+    if (seq == nullptr) {
+      PyErr_Clear();
+    } else {
+      Py_ssize_t n = PySequence_Fast_GET_SIZE(seq);
+      for (Py_ssize_t i = 0; i < n; i++) {
+        PyObject *pypt = PySequence_Fast_GET_ITEM(seq, i);
+        double x, y;
+        if (python_vectorval(pypt, 2, 2, &x, &y, nullptr, nullptr, nullptr) == 0) {
+          result.vertices.push_back(Vector2d(x, y));
+        }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/pyopenscad.cc:702]: Fixed in 221aa7bf7. Profile callback results now only parse entries that are coordinate sequences, preventing flat scalar lists from becoming broadcast points.

File: src/python/pyconversion.cc [hunk 27]
```diff
@@ -254,65 +311,100 @@ PyObject *python_fromvector(const Vector3d vec)
...
-      PyErr_SetString(PyExc_TypeError, "must specify at least 1 path when specified");
+  if (pypaths == nullptr) return result;
+  if (!python_is_sequence(pypaths)) {
+    PyErr_SetString(PyExc_TypeError, "Polygon paths must be a sequence of index sequences");
+    return result;
+  }
+  PyObject *seq = PySequence_Fast(pypaths, "expected a list of paths");
+  if (seq == nullptr) {
+    return result;
+  }
```
[the remark this answers is not in the document]
Comment (nomike) [reply src/python/pyconversion.cc:357]: Fixed in 8d00b35db. The polygon points/paths setter now converts into temporaries, checks PyErr_Occurred(), and only mutates the node on success. Echo coverage verifies invalid setter inputs propagate TypeError.

File: libraries/python/stubs/_openscad-stubs/__init__.pyi [hunk 28]
```diff
@@ -113,7 +116,7 @@ class PyOpenSCAD:
     """4x4 transformation matrix representing the object's origin.
     Initialized as identity matrix."""
 
-    def translate(self, v: Vector3) -> "PyOpenSCAD":
+    def translate(self, v: Vector3| "npt.NDArray[np.float64]") -> "PyOpenSCAD":
```
[the remark this answers is not in the document]
Comment (nomike) [reply libraries/python/stubs/_openscad-stubs/__init__.pyi:119]: Fixed in 8d00b35db. Vector and matrix aliases, plus remaining coordinate parameters, now use Sequence types so tuple inputs accepted at runtime are also accepted by type checkers.

File: tests/test_python_conversions.py [hunk 29]
```diff
@@ -222,6 +255,10 @@ def main() -> int:
     # pythonscad routes Python print() output to stderr, so search both.
     combined = (proc.stdout or "") + (proc.stderr or "")
 
+    if "SKIP" in combined:
+        print("SKIP: numpy not available")
+        return 0
```
[the remark this answers is not in the document]
Comment (nomike) [reply tests/test_python_conversions.py]: Fixed in 8d00b35db. Skip detection now only matches output lines beginning with SKIP:, avoiding accidental substring matches.

File: libraries/python/pythonscad/__init__.py [hunk 30]
```diff
@@ -24,6 +24,13 @@
     ChildRef,
     Openscad,
 )
+from ._vectors import (  # noqa: F401
+    HAS_NUMPY,
+    Matrix4x4,
+    Vector1,
+    Vector2,
+    Vector3,
+)
```
Comment (gsohler) [inline libraries/python/pythonscad/__init__.py:34]: Questioning myself, if these types interact with the pythonscad.vector() function

**Review by gsohler — APPROVED** [review]
Great Addition,
plenty of tests.
Look forward to people using lots of Numpy calculations to establish models
